main/src/ca/uwaterloo/flix/language/ast/SyntaxTree.scala

```diff
@@ -0,0 +1,504 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.ast
+
+import ca.uwaterloo.flix.language.errors.Parser2Error
+
+
+/**
+ * A Syntax tree representing the __syntax__ in a source file.
+ * The tree is built of nodes that hold a [[TreeKind]] and zero or more children.
+ * Each child is either a [[Token]] or another node.
+ * [[SyntaxTree]] is intentionally kept very flexible in definition, so it can encode faulty syntax.
+ * Beware that this means that it gives few guarantees.
+ * There is no guarantee on the amount or presence of children of a certain kind for instance.
+ */
+object SyntaxTree {
+
+  /**
+   * A node in a [[SyntaxTree]]
+   *
+   * @param kind     The kind of the node.
+   * @param loc      The location that the node spans in the source file.
+   * @param children The children of the node.
+   */
+  case class Tree(kind: TreeKind, var loc: SourceLocation, var children: Array[Child]) {
+    /**
+     * Utility function that computes a textual representation of a [[SyntaxTree.Tree]].
+     * Meant for debugging use.
+     */
+    def toDebugString(nesting: Int = 1): String = {
+      val kindName = kind.debug_name match {
+        case Some(name) => s"$name.$kind"
+        case None => s"$kind"
+      }
+
+      s"$kindName (${loc.beginLine}, ${loc.beginCol}) -> (${loc.endLine}, ${loc.endCol}) ${
+        children.map {
+          case Child.Token(token) => s"\n${"  " * nesting}'${token.text}'"
+          case Child.Tree(tree) => s"\n${"  " * nesting}${tree.toDebugString(nesting + 1)}"
+        }.mkString("")
+      }"
+    }
+  }
+
+  sealed trait Child
+
+  /**
+   * A child in a [[SyntaxTree]].
+   * It holds a [[Token]] or another [[SyntaxTree.Tree]].
+   */
+  object Child {
+    /**
+     * A [[SyntaxTree]] child holding a [[Token]].
+     */
+    case class Token(token: ca.uwaterloo.flix.language.ast.Token) extends Child
+
+    /**
+     * A [[SyntaxTree]] child holding a nested [[SyntaxTree.Tree]]
+     */
+    case class Tree(tree: SyntaxTree.Tree) extends Child
+  }
+
+
+  sealed trait TreeKind {
+    def debug_name: Option[String] = None
+  }
+
+  /**
+   * Different kinds of syntax nodes in a [[SyntaxTree]].
+   * The only error kind that holds data is the special [[TreeKind.ErrorTree]].
+   */
+  object TreeKind {
+    /**
+     * A special error kind wrapping a [[Parser2Error]].
+     */
+    case class ErrorTree(error: Parser2Error) extends TreeKind
+
+    case object Annotations extends TreeKind
+
+    case object Argument extends TreeKind
+
+    case object ArgumentNamed extends TreeKind
+
+    case object Arguments extends TreeKind
+
+    case object Case extends TreeKind
+
+    case object Comments extends TreeKind
```

Comment:
Plural vs singular? How to understand? Is it more than one comment.

Reply:
Ah, maybe its LineComments?

Reply:
So it's both line and block comments. The intention is to consume as many subsequent comments into the same tree node of kind `Comments`.

Note that it would also be completely valid to just have a single tree node per comment. I think it wouldn't change much really, since the Weeder just ignores comments.